Source/WebCore/page/Navigation.cpp

```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
     }
 }
 
+static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(Document& document)
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    RefPtr deferredPromise = DeferredPromise::create(globalObject);
+    Ref domPromise = DOMPromise::create(globalObject, *JSC::jsCast<JSC::JSPromise*>(deferredPromise->promise()));
+    return { WTFMove(domPromise), deferredPromise.releaseNonNull() };
+}
+
+// https://webidl.spec.whatwg.org/#wait-for-all
+static void waitForAllPromises(Vector<RefPtr<DOMPromise>>& promises, std::function<void()> fulfilledCallback, std::function<void(JSC::JSValue)> rejectionCallback)
+{
+    auto remainingPromises = std::make_shared<size_t>(promises.size());
+    auto rejected = std::make_shared<bool>(false);
+
+    for (const auto& promise : promises) {
+        // At any point between promises the frame could have been detached.
+        if (promise->isSuspended())
```

Comment:
So if all promises are suspended we will never call either callback.  Is that what we want?  Can they become unsuspended?

Reply:
I'm not certain its possible and I'm not sure what to do in that case.

It may not be possible, like `whenSettled`'s callback probably can't be called from JS when JS is already suspended. These codepaths have only been hit when the frame is actively destroyed during the NavigateEvent.

Reply:
I did try calling rejection in this case also but no test case covered it. I could maybe write one expecting an error after detach to see if that edge-case matters.